_posts/2020-01-23-scatter-gather-pt2.md

```diff
@@ -0,0 +1,142 @@
+* For persistence options using optimistic concurrency control, the endpoint can run into many concurrency violation exceptions. These exceptions will force the losing transactions to retry, potentially causing large amounts of retries. At some point, messages might even be moved to the error queue due to exceding the configured retry limits.
+
+If you'd have to chose between one of the options, you'd probably prefer the pessimistic locking approach. The throughput limitations are probably less severe than the impact of multiple retries and are way better than dealing with messages in the error queue. Luckily, most of the supported Saga persistence options support pessimistic locking for sagas by now (the MongoDB and ServiceFabric persistence packages just recently switched to pessimistic locking). Particular Software also vastly improved the official [documentation on saga concurrency](https://docs.particular.net/nservicebus/sagas/concurrency), describing different concurrency problems and approaches to solve those, a highly recommended read.
+
+In most cases, using a persistence option with pessimistic locking capabilities will prevent major problems for you, at the cost of slower throughput at some times. However, there is another approach which doesn't impact the endpoint throughput and also won't cause concurrency exceptions and retries. It's a lot more complex than just using pessimistic locking but sometimes the throughput limitations can be an important factor or you might rely on a persistence not support pessimistic locking. Here's a rough overview how this approach will work:
+
+1. Instead of aggregating the results in the saga directly, a simple message handler will receive the response message type.
+2. The handler will store the response data in an append-only storage which allows extremely fast inserts without concurrency violations.
+3. A separate process will repeatedly check if all expected responses arrived and will continue the business process managed by the saga.
+4. The saga will aggregate all the response data.
+
+Let's look at some sample code:
+
+### Scattering
+
+The `ScatterGatherSaga` will dispatch a large amount of messages (e.g. 500) when the saga is started:
+
+```
+class ScatterGatherSaga : Saga<ScatterGatherSagaData>, 
+    IAmStartedByMessages<StartSagaCommand>
+{
+    public async Task Handle(StartSagaCommand message, IMessageHandlerContext context)
+    {
+        Data.NumberOfMessages = message.NumberOfMessages;
+
+        for (int i = 0; i < message.NumberOfMessages; i++)
+        {
+            await context.Send(new RequestResponseCommand()
+            {
+                BatchId = message.Id,
+                RequestId = i
+            });
+        }
+    }
+
+    // mapping omitted
+}
+
+```
+
+Another endpoint will respond to each `RequestResponseCommand` with a `ResponseMessage`. Instead of handling those directly as part of the saga, we implement a stateless message handler:
```

Comment:
```suggestion
Another endpoint will respond to each `RequestResponseCommand` with a `ResponseMessage`. Instead of handling those directly as part of the saga, we implement a stateless message handler.
```

Is it worth mentioning that `Reply` cannot be used as the reply would go to the saga?